tools/viewgar.1.in
.TH viewgar 1 "January, 2000" "GiNaC @VERSION@" "The GiNaC Group"
.SH NAME
viewgar \- GiNaC archive file viewer
.SH SYNPOSIS
.B viewgar
[\-d]
.RI [ file\&... ]
.SH DESCRIPTION
.B viewgar
displays the contents of GiNaC archive files. By default it will print the
archived expressions in standard mathematical notation. If given the
.B "\-d"
option it will output a raw dump of the archive contents).
.SH OPTIONS
.TP
.B \-d
print raw dump of archive instead of formatted expressions
.SH AUTHOR
.TP
The GiNaC maintainers <https://www.ginac.de/>.
.SH SEE ALSO
GiNaC Tutorial \- An open framework for symbolic computation within the
C++ programming language
.SH COPYRIGHT
Copyright \(co 1999-2022 Johannes Gutenberg Universit\(:at Mainz, Germany

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301,
USA.


NEWS
This file records noteworthy changes.

1.8.2 (1 January 2022)
* Fix elusive bug in comparing relational objects.
* Ensure modular_form_kernel::series() includes an Order term.

1.8.1 (9 August 2021)
* Add method relational::canonical() and improve conversion of relational to
  Boolean (it now works on many simple symbolic cases).
* Improve normalization of negative exponents.
* Fix indexing multiply referenced objects with ex::operator[].
* Make functions evalf() their arguments before doing own evalf().
* Fix bugs in H_evalf() and in evaluation of iterated integrals.
* Several portability improvements and compiler warning fixes.

1.8.0 (14 October 2020)
* New routines for the numerical evaluation of iterated integrals like
  elliptic multiple polylogarithms or iterated integrals of modular forms.
* Stronger normalization methods for expressions with powers. Where this
  is safe, GiNaC now tries replacing various kinds of powers in order to
  find and cancel common subexpressions.
* Improved CMake build.

1.7.11 (3 August 2020)
* Fix elusive bugs in factor() and in expand().

1.7.10 (28 June 2020)
* Fix collect_common_factors() for hidden zero arguments.
* Fix build on modern systems (C++20 and Python3).

1.7.9 (11 April 2020)
* Fix unarchiving of overloaded functions.
* Fix MinGW64 build.

1.7.8 (7 October 2019)
* Fix pseries::evalf(), broken since 1.7.0.
* Fix a corner-case bug in H_evalf().

1.7.7 (22 September 2019)
* Fix unarchiving empty containers.
* Remove obsolete exhashmap implementation.
* Fix issues with zeta computation.
* Build with Python3.

1.7.6 (22 April 2019)
* Fix a bug in mul::series() about asymptotic order term.
* Fix cl_N code generation for small negative integers.
* Fix multiple polylog computation related to trailing zeros and user-specified
  signs for small imaginary parts.
* Fix CMake build with Python3.

1.7.5 (17 February 2019)
* Add Markowitz-ordered Gaussian eliminatin.
* Avoid un-normal zeros in matrix::solve().
* Accept exprseq arguments in lsolve(eqns, vars).
* Add overload of matrix::rank(solve_algo).
* Fix case in G_do_hoelder with real x values which are not cl_R.
* Fix an elusive bug in expairseq/add ctor, introduced in 1.7.0.

1.7.4 (19 February 2018)
* Fix ABI compatibility with 1.7.[0-2] (broken unintentionally in 1.7.3).

1.7.3 (1 February 2018)
* Speed up some corner cases of polynomial factorization.
* Speed up some corner cases of gcd computation.
* Speed up factorization of pre-factored polynomials.
* Make matrix::solve() work with hidden (non-normal) zeros.
* Fix bug in evaluation of real and imaginary parts of powers.
* Fix some LaTeX output, add new function symbol::get_TeX_name().

1.7.2 (10 January 2017)
* Fix memory leak in operators returning relationals, introduced in 1.7.0.
* Make macro GINAC_DECLARE_REGISTERED_CLASS() work outside namespace GiNaC.
* Fix some build issues.

1.7.1 (2 October 2016)
* Fix crash in basic::subs(), visible in classes which don't override it.
* Fix convergence failure in Li_projection.
* Fix cliffor_bar() and clifford_star().
* Improve performance of clifford_to_lst().
* Fix compilation issues, reduce compiler warnings.

1.7.0 (30 April 2016)
* Make use of C++11 features:
  - Use new language features 'auto', 'nullptr', 'constexpr', range-based for
    loops, the contextual keyword 'override', static 'noexcept' operator, etc.
  - Add initializer list ctors for container<>, lst and nested initializer
    lists ctor for class matrix.
  - Add many rvalue reference overloads for ctors from std containers. (Note
    that whether they really speed up things depends on whether evaluation
    leaves the container unchanged or not.)
  - Rvalue reference overloads for ctors from std containers made it possible
    to remove the manual 'discardable' hack and std::auto_ptr<T> for good.
  - New dynallocate<T>(args...) helper functions obsolete manual allocation
    on the heap and setting the dynallocated status flag.
  - Got rid of C++98ish algorithm features (bind2nd, etc.) and replaced them
    with lambdas.
* Make add::eval(), mul::eval() more careful so they return an evaluated,
  canonical expression without compromise.
* Remove 'level' argument of all .eval() methods. This has become possible by
  ensuring that add::eval() and mul::eval() evaluate their own structure
  without compromise (this was not guaranteed before). If all ex are evaluated,
  any object which is an aggregate of such ex will only ever have to
  evaluate itself, obsoleting the need for recursing.
* Remove 'level' arguments of all .evalf() and .normal() methods. They were
  modeled after the .eval() methods.
* Make relational::rhs() and lhs() nonvirtual.
* Add support for power::info(info_flags::real).
* Add fderivative::derivatives() method.
* Remove broken info_flags::algebraic.
* Remove unmaintained support for Cint.
* Remove unfinished andcode for hash table-based expairseq.
* Remove deprecated ex::to_rational(lst) and ex::to_polynomial(lst).
* ginac-excompiler installs in $LIBEXECDIR now, not in $BINDIR.
* Many minor performance enhancements.

1.6.7 (29 February 2016)
* Fix elusive bug in sqrfree(), normal(), numer(), and denom().

1.6.6 (20 December 2015)
* Fix elusive bug in add::eval().
* Several minor performance enhancements.

1.6.5 (27 July 2015)
* Fix pow(+(...),2).expand() where the sum contained roots of expressions.
* Speed-up pow(+(...),n).expand() where n > 3.
* Fix a crash in PolyLogs of rational numbers.

1.6.4 (8 May 2015)
* Fix pow(+(...),2).expand() where the sum contained roots of rationals.
* Various improvements for info(info_flags::nonnegative).
* Make power::expand() (x*p)^c -> x^c * p^c, if p>0.
* Fix normal() where nested functions occur multiple times.
* Add optional explicit derivatives to functions and use it in abs(),
  imag_part(), real_part().
* Evaluate abs(z)^(2n) -> z^n * conjugate(z)^n.

1.6.3 (27 November 2014)
* New expand_options::expand_transcendental. GiNaC will apply additional

tools/viewgar.cpp
/** @file viewgar.cpp
 *
 *  GiNaC archive file viewer. */

/*
 *  GiNaC Copyright (C) 1999-2022 Johannes Gutenberg University Mainz, Germany
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include "ginac.h"
using namespace GiNaC;

#include <cstdlib>
#include <fstream>
#include <iostream>
#include <stdexcept>
using namespace std;

int main(int argc, char **argv)
{
	if (argc < 2) {
		cerr << "Usage: " << argv[0] << " [-d] file..." << endl;
		exit(1);
	}
	--argc; ++argv;

	bool dump_mode = false;
	try {
		lst l;
		while (argc) {
			if (strcmp(*argv, "-d") == 0) {
				dump_mode = true;
				--argc; ++argv;
			}
			std::ifstream f(*argv, std::ios_base::binary);
			archive ar;
			f >> ar;
			if (dump_mode) {
				ar.printraw(std::cout);
				std::cout << std::endl;
			} else {
				for (unsigned int i=0; i<ar.num_expressions(); ++i) {
					std::string name;
					ex e = ar.unarchive_ex(l, name, i);
					std::cout << name << " = " << e << std::endl;
				}
			}
			--argc; ++argv;
		}
	} catch (std::exception &e) {
		std::cerr << *argv << ": " << e.what() << std::endl;
	}
}
